### DISCUSSION TITLE
Beta Roadmap

### DISCUSSION THREAD

**Discussion by chaance** [radix-ui/primitives#375] [Ideas]
Copied from https://github.com/radix-ui/primitives/issues/371 so we can close that and create individual issues for action items.

@colmtuite @benoitgrelard @jjenzz 

### General

- [ ] Ensure touch support is good across all primitives.
- [ ] Test all primitives rigorously with Stitches and SC, raising issues along the way.
- [ ] Complete any unfinished features (like the positioning logic for DropdownMenu raised in the design tool).

### Docs

- [x] Add UI to docs for cycling through docs versions. (Pedro)
- [ ] Add new top-level "Composition" section, mentioning `forwardRef` etc. (Chance)
- [ ] Add examples to relevant primitives showing how to wrap them in good consumer APIs. (Pedro)
- [ ] Add tutorial videos for each primitive. (Pedro)

### New Primitives

Loosely in order of priority.

- [x] Submenu (Andy)
- [ ] Select (Benoît)
- [ ] Menubar (Andy)
- [x] ToggleButtonGroup (Chance)
- [ ] Toast (Jenna)
- [x] HoverCard (Andy)
- [ ] Carousel (Jenna)
- [x] Toolbar

**Comment by chaance** [comment]
> Complete any unfinished features (like the positioning logic for DropdownMenu raised in the design tool).

@colmtuite @benoitgrelard do either of you know what specific issue this is regarding? Just wanted to make sure we are tracking it.

Are there any other unfinished features we need to consider for a beta release?

**Comment by benoitgrelard** [comment]
Yeah I believe that's to do with the fact that the menu can overflow the view now and we'd like to constrain its height (at least in some situations). I wanted to do that before xmas but didn't have time to get to it.

**Reply by DaniGuardiola** [reply]
Don't wanna be the guy who replies "+1" but...

+1

Just kidding, thanks for your work, I'll wait patiently for this :)

**Comment by acc-nicholas** [comment]
@benoitgrelard hey, my team was asking about the status of `Select` and when it is planned for release in some form or another. Do you have a general timeline or estimate that i can relay back to my team? Ballpark is fine.

**Reply by schonert** [reply]
Also quite curious about this

**Reply by laustdeleuran** [reply]
+1 👀

**Reply by colmtuite** [reply]
Select is next on our list. The best estimate I can provide is ~5–6 weeks from now. But that's very loose. We'll get it done asap!

**Reply by jillesmc** [reply]
Hi @colmtuite and @benoitgrelard 
Is the `Select` component already a work in progress? 😃

**Reply by colmtuite** [reply]
We'll get it done asap.

**Reply by eespaillat** [reply]
hi sorry for to bother but wondering if there is a new estimate on when it might ship? Need one but wondering if it makes sense to wait for the release.

**Reply by jjenzz** [reply]
@benoitgrelard is working on this at the moment but is away on holiday for the next few weeks so it'll most likely be early September now.

**Reply by gazpachu** [reply]
@benoitgrelard any update on this? Let me know if you need some help.

**Reply by creeefs** [reply]
Also, eagerly awaiting the `Select` component! 🙏

**Reply by noahsark769** [reply]
Also interested in trying out the Select component! Happy to help with this if I can

**Reply by acc-nicholas** [reply]
@benoitgrelard Any update on `Select`? Its been over 6 months and i understand this one is particularly complex, but a time estimate can inform my teams plans.

**Reply by benoitgrelard** [reply]
We're in the final phases of testing, so hopefully next month 🙂

**Reply by josias-r** [reply]
@benoitgrelard Awesome! Is there any beta/alpha available that can already be tested out? 😄

**Reply by benoitgrelard** [reply]
Not yet, but we may release a candidate at some point to test out.

**Comment by kawpii** [comment]
@benoitgrelard @jjenzz @andy-hook @chaance @peduarte 

Great library! Thank you guys! Have more components than HeadlessUI. I really like headless, accessible components.

----------
Some feedback:

I wish team focus on doing implementations of components which could not be implemented using CSS, looking at current docs following components could be easily implemented with CSS

- Avatar
- Label
- AspectRatio
- Separator

See this [comment](https://github.com/mui-org/material-ui/issues/27170#issue-939062130) from MUI to understand what I meant. You will see some components are marked as
>  It seems to be a UI only problem.

Some components I like to see implemented

Common :
- Drawer
- Tags / Chips
- Pagination
- DatePicker
- ~~Skeleton~~
- Autocomplete
- [Notification](https://ant.design/components/notification/)

Good to have :
- Circular Progress
- [Rating](https://ant.design/components/rate/)
- [Timeline](https://ant.design/components/timeline/)
- [Tree](https://ant.design/components/tree/)
- [TreeSelect](https://ant.design/components/tree-select/)
- [Cascader](https://ant.design/components/cascader/)


-----------
Again thank you guys! keep up the good work


Update: removed skeleton

**Reply by sanduluca** [reply]
I am also interested in the Cascader and Timeline components

**Comment by colmtuite** [comment]
Thanks for the feedback @kaushalyap.

Our Avatar, Label, and Separator include aria features beyond CSS, so that's why we implemented them. They're relatively simple anyway, so they haven't drained much time.

Most of the components you've listed are on our radar. You could do Chip and Skeleton with CSS, if that helps. We have an example of Skeleton in our DS. Cascader and TreeSelect are a bit odd, I'd say those are the only two we won't get around to any time soon. But we'll get around to the others asap.

Thanks again for the feedback, we'll keep at it!

Note: in future, please don't @ the entire team in posts like this, it adds to their pile of notifications unnecessarily. Thanks ❤️

**Reply by kawpii** [reply]
> You could do Chip and Skeleton with CSS

Chip also have accessibility aspect when there is an icon (ex: close)

> Note: in future, please don't @ the entire team in posts like this, it adds to their pile of notifications unnecessarily.

Okay I'll keep that in mind but thanks should be given to all the members not to a single member.

> Cascader and TreeSelect are a bit odd

Yes they are not frequent but cascader useful in app like Google Docs, TreeSelect also in a case of complex data intensive application.

**Comment by davemooreuws** [comment]
Great library, it would be great to have search capabilities on the radix and stitches doc sites. Algolia doc search is a good option.

**Comment by kungpaogao** [comment]
Is it possible to try out beta components? I would love to use `Carousel` to create something like what's on the Radix homepage!

**Reply by jjenzz** [reply]
The `beta` components are all at https://radix-ui.com (the project is in beta). Anything that is not there is too unstable to share atm I'm afraid. 

If I recall, the `Carousel` on the homepage had some hardcoded features to get it working for that particular use-case so wouldn't work for everyone.

**Reply by wiesson** [reply]
I'm also super interested in the `Carousel` component :)

**Comment by JuanM04** [comment]
Amazing library, I love it. I've seen previews of the Select component and it looks fantastic!

Btw, do you have plans on doing a Combobox component?

**Reply by benoitgrelard** [reply]
It's on the list yes, that's a big one though, no precise plans yet.

**Reply by xixixao** [reply]
Where could I see a preview of select, or could you share the design spec? Is it a "typeahead" component?

**Reply by benoitgrelard** [reply]
You can think of it as basically the same thing as the native MacOS select, but styleable.

**Reply by lukebennett88** [reply]
> Where could I see a preview of select, or could you share the design spec? Is it a "typeahead" component?

@xixixao I think @JuanM04 was talking about this post: https://www.modulz.app/blog/under-the-spotlight-select

**Reply by colmtuite** [reply]
@lukebennett88 It will be released in a few weeks.

**Reply by lukebennett88** [reply]
@colmtuite that's great news, I can't wait!

**Reply by hollandThomas** [reply]
Fantastic news! 🎉
I'm curious, are there any future plans to support multi-selections? That'd be dope!

**Reply by colmtuite** [reply]
Hopes, but no current plans.

**Reply by kawpii** [reply]
@colmtuite Why no new components after July (according to the top comment edit history)?

**Reply by colmtuite** [reply]
Because components are complex and take a long time to build.

**Comment by p6l-richard** [comment]
I worked on a Stepper component the other day and couldn't help but miss radix' style DX by symply passing a `<Stepper.Next />` or `<Stepper.Previous />` component to wherever I'd like them to be (similar to the `Close` component of the Dialog primitive).

Any plans to build something like this?

For clarification, when I say stepper component I'm talking about something [like this](https://mui.com/material-ui/react-stepper/#linear).
